### ISSUE TITLE
KeyError: 'namespace' when downgrade to stable version 0.20 from 0.21rc3

### ISSUE THREAD

**Issue by smileisak** [zalando-incubator/kopf#180]
## Actual Behavior

When downgrading `kopf` version from `0.21rc` to `0.20` i've got an error when creating, updating and deleting objects.

## Specifications

  - Platform: GCP GKE 
  - Kubernetes version: `v1.13.6-gke.13`
  - Python version: `3.7.4`
  - Python packages installed: *(use `pip freeze --all`)*

```bash
adal==1.2.1
aiofiles==0.4.0
aiohttp==3.5.4
aiojobs==0.2.2
asn1crypto==0.24.0
async-timeout==3.0.1
atomicwrites==1.3.0
attrs==19.1.0
cachetools==3.1.0
certifi==2018.11.29
cffi==1.12.2
chardet==3.0.4
Click==7.0
coloredlogs==10.0
cryptography==2.6.1
decorator==4.4.0
fut==0.4.2
google-api-python-client==1.7.8
google-auth==1.6.3
google-auth-httplib2==0.0.3
httplib2==0.12.1
humanfriendly==4.18
idna==2.8
iso8601==0.1.12
isort==4.3.17
jsonpath-ng==1.4.3
kopf==0.20
kubernetes==8.0.1
more-itertools==6.0.0
multidict==4.5.2
oauth2client==4.1.3
oauthlib==3.0.1
pip==19.1.1
pluggy==0.9.0
ply==3.11
protobuf==3.6.1
py==1.8.0
pyasn1==0.4.5
pyasn1-modules==0.2.4
pycparser==2.19
PyJWT==1.7.1
pykube-ng==0.28
pyotp==2.2.7
pytest==4.3.0
python-anticaptcha==0.3.1
python-dateutil==2.8.0
pytz==2019.1
PyYAML==3.13
requests==2.21.0
requests-oauthlib==1.2.0
rsa==4.0
setuptools==40.6.3
six==1.12.0
tzlocal==2.0.0
uritemplate==3.0.0
urllib3==1.24.1
urllib3-mock==0.3.3
websocket-client==0.55.0
wheel==0.32.3
yarl==1.3.0
```

---
## Error:

```bash
2019-08-10 11:13:15 provisioner-j2c-operator-f45b6c7c6-ffq2l kopf.clients.events[1] WARNING Failed to post an event. Ignoring and continuing. Error: HTTPError('Event "kopf-event-2xt2t" is invalid: involvedObject.namespace: Invalid value: "": does not match event.namespace'). Event: type='Error', reason='Logging', message='Handler \'create_or_update_fn\' failed with an exception. Will retry.\nTraceback (most recent call last):\n  File "/app/env/lib/python3.7/site-packages/kopf/reactor/handling.py", line 387, in _execute\n    lifecycle=lifecycle,  # just a default for the sub-handlers, not used directly.\n  File "/app/env/lib/python3.7/site-packages/kopf/reactor/handling.py", line 478, in _call_handler\n    **kwargs,\n  File "/app/env/lib/python3.7/site-packages/kopf/reactor/invocation.py", line 79, in invoke\n    result = await loop....ssage=\'Start handling create/update action\')\n  File "/app/env/lib/python3.7/site-packages/kopf/engines/posting.py", line 79, in info\n    event(obj, type=\'Normal\', reason=reason, message=message)\n  File "/app/env/lib/python3.7/site-packages/kopf/engines/posting.py", line 72, in event\n    ref = hierarchies.build_object_reference(obj)\n  File "/app/env/lib/python3.7/site-packages/kopf/structs/hierarchies.py", line 16, in build_object_reference\n    namespace=body[\'metadata\'][\'namespace\'],\nKeyError: \'namespace\''.
2019-08-10 11:14:15 provisioner-j2c-operator-f45b6c7c6-ffq2l kopf.objects[1] ERROR Handler 'create_or_update_fn' failed with an exception. Will retry.
Traceback (most recent call last):
  File "/app/env/lib/python3.7/site-packages/kopf/reactor/handling.py", line 387, in _execute
    lifecycle=lifecycle,  # just a default for the sub-handlers, not used directly.
  File "/app/env/lib/python3.7/site-packages/kopf/reactor/handling.py", line 478, in _call_handler
    **kwargs,
  File "/app/env/lib/python3.7/site-packages/kopf/reactor/invocation.py", line 79, in invoke
    result = await loop.run_in_executor(config.WorkersConfig.get_syn_executor(), real_fn)
  File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/app/operators/run.py", line 26, in create_or_update_fn
    kopf.info(body, reason='Info', message='Start handling create/update action')
  File "/app/env/lib/python3.7/site-packages/kopf/engines/posting.py", line 79, in info
    event(obj, type='Normal', reason=reason, message=message)
  File "/app/env/lib/python3.7/site-packages/kopf/engines/posting.py", line 72, in event
    ref = hierarchies.build_object_reference(obj)
  File "/app/env/lib/python3.7/site-packages/kopf/structs/hierarchies.py", line 16, in build_object_reference
    namespace=body['metadata']['namespace'],
KeyError: 'namespace'
```

**Comment by nolar** [comment]
@smileisak Is it for cluster-scoped objects or namespaced objects? Did it actually happen after the downgrade, or just in general (no matter the previously installed versions)?

Maybe a duplicate of #164. Should be solved in #165 — which is now available as [0.21rc1](https://github.com/zalando-incubator/kopf/releases/tag/0.21rc1) the earliest.

**Comment by smileisak** [comment]
@nolar is a cluster-scoped object. In the begining i've created a namespaced CRD that i have changed it to cluster-scoped. Now it works only when using version `0.21rc`.

**Comment by nolar** [comment]
I assume that the issue is solved with 0.21 released (since 0.21rc1 actually). If it reappears, feel free to re-open or just write a message here.
